R/pmcmc.R

```diff
@@ -0,0 +1,609 @@
+
+  coda_res <- coda::as.mcmc(res)
+  rejection_rate <- coda::rejectionRate(coda_res)
+  ess <- coda::effectiveSize(coda_res)
+
+  out <- list("results" = as.data.frame(res),
+              "acceptance_rate" = 1 - rejection_rate,
+              "ess" = ess)
+
+  if (output_proposals) {
+    proposals <- as.data.frame(proposals)
+    out$proposals <- proposals
+  }
+
+ class(out) <- "mcstate_pmcmc"
+ out
+}
+
+calc_lprior <- function(pars, pars_lprior) {
+  lprior <- 0
+  for (par in names(pars)) {
+      lprior <- lprior + pars_lprior[[par]](pars)
+  }
+  names(lprior) <- NULL
+  lprior
+}
+
+propose_parameters <- function(pars, proposal_kernel, pars_discrete,
+                               pars_min, pars_max) {
+
+  ## proposed jumps are normal with mean pars and sd as input for parameter
+  jumps <- pars + drop(rmvnorm(n = 1, sigma = proposal_kernel))
+
+  # discretise if necessary
+  jumps[pars_discrete] <- round(jumps[pars_discrete])
+  # reflect proposal if it exceeds upper or lower parameter boundary
+  jumps <- reflect_proposal(x = jumps,
+                          floor = pars_min,
+                          cap = pars_max)
+  jumps
+}
+
+## create function to reflect proposal boundaries at pars_min and pars_max
+# this ensures the proposal is symetrical and we can simplify the MH step
+reflect_proposal <- function(x, floor, cap) {
+  interval <- cap - floor
+  abs((x + interval - floor) %% (2 * interval) - interval) + floor
+}
+
+# Check input data.frame
+mcmc_validate_range <- function(range) {
+  assert_is(range, "data.frame")
+  msg <- setdiff(c("name", "init", "min", "max", "discrete", "target"),
+                   names(range))
+  if (length(msg) > 0L) {
+    stop("Missing columns from 'mcmc_range': ",
+           paste(squote(msg), collapse = ", "))
+  }
+
+  if (anyDuplicated(range$name)) {
+    stop("Duplicate 'name' entries not allowed in 'mcmc_range'")
+  }
+
+  targets <- c("step_start", "model_data", "pars_compare")
+  err <- setdiff(range$target, targets)
+  if (length(err) > 0L) {
+    stop(sprintf("Invalid target %s: must be one of %s",
+                 paste(squote(err), collapse = ", "),
+                 paste(squote(targets), collapse = ", ")))
+  }
+
+  index <- lapply(targets, function(t) which(range$target == t))
+  names(index) <- targets
+  if (length(index$step_start) > 1L) {
+    stop("At most one target may be 'step_start'")
+  }
+
+  if (!is.logical(range$discrete)) {
+    stop("'discrete' entries must be TRUE or FALSE")
+  }
+  if (!is.numeric(range$init)) {
+    stop("'init' entries must be numeric")
+  }
+  if (!is.numeric(range$min)) {
+    stop("'min' entries must be numeric")
+  }
+  if (!is.numeric(range$max)) {
+    stop("'max' entries must be numeric")
+  }
+
+  if (any(range$init < range$min | range$init > range$max)) {
+    stop("initial parameters are outside of specified range")
+  }
+
+  list(range = range,
+       index = index)
+}
+
+##' @title create a master chain from a pmcmc_list object
```

Comment:
What is a master chain? As this is an exported function can it get defined here?